{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd272da8",
   "metadata": {},
   "source": [
    "****Randomized Motif Search algorithm****"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a94246de",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"Download the sequence of 10 genes of DosR dataset.\"\n",
    "\n",
    "import wget\n",
    "\n",
    "DosR = wget.download('https://bioinformaticsalgorithms.com/data/challengedatasets/DosR.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "75fb5b5e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCGCCCCGCCCGGACAGCCATGCGCTAACCCTGGCTTCGATGGCGCCGGCTCAGTTAGGGCCGGAAGTCCCCAATGTGGCAGACCTTTCGCCCCTGGCGGACGAATGACCCCAGTGGCCGGGACTTCAGGCCCTATCGGAGGGCTCCGGCGCGGTGGTCGGATTTGTCTGTGGAGGTTACACCCCAATCGCAAGGATGCATTATGACCAGCGAGCTGAGCCTGGTCGCCACTGGAAAGGGGAGCAACATC',\n",
       " 'CCGATCGGCATCACTATCGGTCCTGCGGCCGCCCATAGCGCTATATCCGGCTGGTGAAATCAATTGACAACCTTCGACTTTGAGGTGGCCTACGGCGAGGACAAGCCAGGCAAGCCAGCTGCCTCAACGCGCGCCAGTACGGGTCCATCGACCCGCGGCCCACGGGTCAAACGACCCTAGTGTTCGCTACGACGTGGTCGTACCTTCGGCAGCAGATCAGCAATAGCACCCCGACTCGAGGAGGATCCCG',\n",
       " 'ACCGTCGATGTGCCCGGTCGCGCCGCGTCCACCTCGGTCATCGACCCCACGATGAGGACGCCATCGGCCGCGACCAAGCCCCGTGAAACTCTGACGGCGTGCTGGCCGGGCTGCGGCACCTGATCACCTTAGGGCACTTGGGCCACCACAACGGGCCGCCGGTCTCGACAGTGGCCACCACCACACAGGTGACTTCCGGCGGGACGTAAGTCCCTAACGCGTCGTTCCGCACGCGGTTAGCTTTGCTGCC',\n",
       " 'GGGTCAGGTATATTTATCGCACACTTGGGCACATGACACACAAGCGCCAGAATCCCGGACCGAACCGAGCACCGTGGGTGGGCAGCCTCCATACAGCGATGACCTGATCGATCATCGGCCAGGGCGCCGGGCTTCCAACCGTGGCCGTCTCAGTACCCAGCCTCATTGACCCTTCGACGCATCCACTGCGCGTAAGTCGGCTCAACCCTTTCAAACCGCTGGATTACCGACCGCAGAAAGGGGGCAGGAC',\n",
       " 'GTAGGTCAAACCGGGTGTACATACCCGCTCAATCGCCCAGCACTTCGGGCAGATCACCGGGTTTCCCCGGTATCACCAATACTGCCACCAAACACAGCAGGCGGGAAGGGGCGAAAGTCCCTTATCCGACAATAAAACTTCGCTTGTTCGACGCCCGGTTCACCCGATATGCACGGCGCCCAGCCATTCGTGACCGACGTCCCCAGCCCCAAGGCCGAACGACCCTAGGAGCCACGAGCAATTCACAGCG',\n",
       " 'CCGCTGGCGACGCTGTTCGCCGGCAGCGTGCGTGACGACTTCGAGCTGCCCGACTACACCTGGTGACCACCGCCGACGGGCACCTCTCCGCCAGGTAGGCACGGTTTGTCGCCGGCAATGTGACCTTTGGGCGCGGTCTTGAGGACCTTCGGCCCCACCCACGAGGCCGCCGCCGGCCGATCGTATGACGTGCAATGTACGCCATAGGGTGCGTGTTACGGCGATTACCTGAAGGCGGCGGTGGTCCGGA',\n",
       " 'GGCCAACTGCACCGCGCTCTTGATGACATCGGTGGTCACCATGGTGTCCGGCATGATCAACCTCCGCTGTTCGATATCACCCCGATCTTTCTGAACGGCGGTTGGCAGACAACAGGGTCAATGGTCCCCAAGTGGATCACCGACGGGCGCGGACAAATGGCCCGCGCTTCGGGGACTTCTGTCCCTAGCCCTGGCCACGATGGGCTGGTCGGATCAAAGGCATCCGTTTCCATCGATTAGGAGGCATCAA',\n",
       " 'GTACATGTCCAGAGCGAGCCTCAGCTTCTGCGCAGCGACGGAAACTGCCACACTCAAAGCCTACTGGGCGCACGTGTGGCAACGAGTCGATCCACACGAAATGCCGCCGTTGGGCCGCGGACTAGCCGAATTTTCCGGGTGGTGACACAGCCCACATTTGGCATGGGACTTTCGGCCCTGTCCGCGTCCGTGTCGGCCAGACAAGCTTTGGGCATTGGCCACAATCGGGCCACAATCGAAAGCCGAGCAG',\n",
       " 'GGCAGCTGTCGGCAACTGTAAGCCATTTCTGGGACTTTGCTGTGAAAAGCTGGGCGATGGTTGTGGACCTGGACGAGCCACCCGTGCGATAGGTGAGATTCATTCTCGCCCTGACGGGTTGCGTCTGTCATCGGTCGATAAGGACTAACGGCCCTCAGGTGGGGACCAACGCCCCTGGGAGATAGCGGTCCCCGCCAGTAACGTACCGCTGAACCGACGGGATGTATCCGCCCCAGCGAAGGAGACGGCG',\n",
       " 'TCAGCACCATGACCGCCTGGCCACCAATCGCCCGTAACAAGCGGGACGTCCGCGACGACGCGTGCGCTAGCGCCGTGGCGGTGACAACGACCAGATATGGTCCGAGCACGCGGGCGAACCTCGTGTTCTGGCCTCGGCCAGTTGTGTAGAGCTCATCGCTGTCATCGAGCGATATCCGACCACTGATCCAAGTCGGGGGCTCTGGGGACCGAAGTCCCCGGGCTCGGAGCTATCGGACCTCACGATCACC']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open('DosR.txt','r') as DosR_seq:\n",
    "    \n",
    "    DosR = []\n",
    "    for i in DosR_seq:\n",
    "        DosR.append(i.rstrip())\n",
    "DosR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4eeb82c4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CGGGACTTCAGGCCC', 'CGGGTCCATCGACCC', 'CGGGACGTAAGTCCC', 'TTGACCCTTCGACGC', 'CGTGACCGACGTCCC', 'GAGGACCTTCGGCCC', 'GGGGACTTCTGTCCC', 'TGGGACTTTCGGCCC', 'GGGGACCAACGCCCC', 'GGGGACCGAAGTCCC']\n",
      "37\n"
     ]
    }
   ],
   "source": [
    " \"\"\"Randomized Motif Search algorithm that takes a list of strings Dna along with integers k and t as input and returns motifs with best (lowest) score.\"\"\"\n",
    "\n",
    "# import the random package here\n",
    "from random import randint\n",
    "# Input:  Positive integers k and t, followed by a list of strings Dna\n",
    "# Output: RandomizedMotifSearch(Dna, k, t)\n",
    "def RandomizedMotifSearch(Dna, k, t):\n",
    "    # insert your code here\n",
    "    M = RandomMotifs(Dna, k, t)\n",
    "    BestMotifs = M\n",
    "    \n",
    "    while True:\n",
    "        Profile = ProfileWithPseudocounts(M)\n",
    "        M = Motifs(Profile, Dna)\n",
    "        if Score(M) < Score(BestMotifs):\n",
    "            BestMotifs = M\n",
    "        else:\n",
    "            return BestMotifs \n",
    "       \n",
    "        \n",
    "        \n",
    "# placed all subroutines needed for Randomized Motif Search algorithm below this line.\n",
    "\n",
    "\"CountWithPseudocounts(Motifs) function\"\n",
    "def CountWithPseudocounts(Motifs):\n",
    "    t = len(Motifs)\n",
    "    k = len(Motifs[0])\n",
    "    \n",
    "    Count = {}\n",
    "    for symbol in 'ACGT':\n",
    "        Count[symbol] = []\n",
    "        for i in range(k):\n",
    "            Count[symbol].append(1)\n",
    "    for n in range(t):\n",
    "        for j in range(k):\n",
    "            symbol = Motifs[n][j]\n",
    "            Count[symbol][j] += 1\n",
    "    return Count\n",
    "\n",
    "\n",
    "\"ProfileWithPseudocounts(Motifs) function\"\n",
    "def ProfileWithPseudocounts(Motifs):\n",
    "    t = len(Motifs)\n",
    "    k = len(Motifs[0])\n",
    "    profile = {} # output variable\n",
    "    # your code here\n",
    "    for key in 'ATGC':\n",
    "        profile[key] = []\n",
    "        for i in range(k):\n",
    "            profile[key].append(1/(t+4)) #count matrix with four more additional counts in each position.\n",
    "    for i in range(t):                    # So, we need to add 4 to t in order to have the correct normalization.\n",
    "        for j in range(k):\n",
    "            key = Motifs[i][j]\n",
    "            profile[key][j] += 1/(t+4)\n",
    "    return profile\n",
    "\n",
    "\n",
    "\"Consensus(Motifs) function\"\n",
    "def Consensus(Motifs):\n",
    "    # insert your code here\n",
    "    k = len(Motifs[0])\n",
    "    count = CountWithPseudocounts(Motifs)\n",
    "    \n",
    "    Consensus = ''\n",
    "    for j in range(k):\n",
    "        m = 0\n",
    "        frequentSymbol = ''\n",
    "        for item in 'ATGC':\n",
    "            if count[item][j] > m:\n",
    "                m = count[item][j]\n",
    "                frequentSymbol = item\n",
    "        Consensus += frequentSymbol\n",
    "    return Consensus\n",
    "\n",
    "\n",
    "\"Score(Motifs) function\"\n",
    "def Score(Motifs):\n",
    "    consensus = Consensus(Motifs)\n",
    "    s = 0\n",
    "    for motif in Motifs:\n",
    "        for item in range(len(motif)):\n",
    "            if motif[item] != consensus[item]:\n",
    "                s += 1\n",
    "    return s\n",
    "\n",
    "\n",
    "\"Probability(Pattern, Profile) function\"\n",
    "def Pr(Pattern, Profile):\n",
    "    t = 1\n",
    "    for i in range(len(Pattern)):\n",
    "        for j in Profile:\n",
    "            if Pattern[i] == j:\n",
    "                t *= Profile[j][i]\n",
    "    return t\n",
    "\n",
    "\n",
    "\"ProfileMostProbablePattern(Text, k, Profile) function\"\n",
    "def ProfileMostProbablePattern(Text, k, Profile):\n",
    "    max_probability = 0\n",
    "    most_probable_pattern = Text[:k]\n",
    "    \n",
    "    for i in range(len(Text) - k + 1):\n",
    "        Pattern = Text[i: i+k]\n",
    "        if Pr(Pattern,Profile) > max_probability:\n",
    "            max_probability = Pr(Pattern,Profile)\n",
    "            most_probable_pattern = Pattern\n",
    "    return most_probable_pattern\n",
    "\n",
    "\"Motifs(Profile,dna) function\"\n",
    "def Motifs(Profile,dna):\n",
    "\n",
    "    k = len(Profile['A'])\n",
    "    Motifs = []\n",
    "    for i in range(0,len(dna)):\n",
    "        km = []\n",
    "        sc = []\n",
    "        for j in range(len(dna[i])-k+1):\n",
    "            km += [dna[i][j:j+k]]\n",
    "        for n in km:\n",
    "            sc += [Pr(n,Profile)]\n",
    "        Motifs += [km[sc.index(max(sc))]]\n",
    "\n",
    "    return Motifs\n",
    "\n",
    "\"RandomMotifs(Dna, k, t) function\"\n",
    "def RandomMotifs(Dna, k, t):\n",
    "    motifs = []\n",
    "    for i in Dna:\n",
    "        s = []\n",
    "        for j in range(len(i) - k+1):\n",
    "            seq = i[j:j+k]\n",
    "            s.append(seq)\n",
    "        rand = randint(1, len(s))\n",
    "        motifs.append(s[rand-1])\n",
    "    return motifs\n",
    "\n",
    "\n",
    "# Copy DosR dataset below.\n",
    "Dna = DosR\n",
    "\n",
    "# set t equal to the number of strings in Dna, k equal to 15, and N equal to 100\n",
    "t = len(Dna)\n",
    "k = 15\n",
    "N = 100\n",
    "\n",
    "# Call RandomizedMotifSearch(Dna, k, t) N times, store the best-scoring set of motifs in a variable called BestMotifs\n",
    "BestMotifs = RandomizedMotifSearch(Dna, k, t)\n",
    "\n",
    "for i in range(N):\n",
    "    M = RandomizedMotifSearch(Dna, k, t)\n",
    "    if Score(M) < Score(BestMotifs):\n",
    "        BestMotifs = M\n",
    "        \n",
    "\n",
    "# Print the BestMotifs variable\n",
    "print(BestMotifs)\n",
    "# Print Score(BestMotifs)\n",
    "print(Score(BestMotifs))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71b9e268",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
